Re-groups and tags enrollments in the program drawer

jkachel opened the pull request:
#### Pre-Flight checklist

- [X] Screenshots and design review for any changes that affect layout or styling
  - [X] Desktop screenshots
  - [X] Mobile width screenshots
- [X] Testing
  - [X] Code is tested
  - [X] Changes have been manually tested

#### What are the relevant tickets?

Fixes #1190 
Fixes #1193 (the code doesn't work in that way anymore so that issue is moot)

#### What's this PR do?

This PR adds grouping of courses in the program drawer according to the program's set requirements. It also resets some of the styling attributes for the enrolled item cards and program drawer to better match the InVision design here: https://projects.invisionapp.com/d/main?origin=v7#/console/21535972/468431707/preview?scrollOffset=0 

To summarize, this:
* Changes the program drawer to display a "Program overview" that consists of a total course count, not enrolled count, and passed course count
* Groups the available courses in the program by their status as an elective or a requirement in the course
* Updates the item cards in the program drawer to display an Enroll message when the user isn't enrolled in the course
* Adds tags (back) to the cards (on both the regular dashboard and in the program drawer):
   * "In progress" for courses whose start date is before now
   * "Enrolled in certificate track" for courses where the learner has a verified enrollment
   * "Course passed" for courses that the learner has a passing grade for 
* Updates the upsell message (either enroll or get certificate) to be formatted in a stacked manner _only if the card is displayed in the program drawer_

This PR also adds a serializer for grades, some helper functions to mark courses in a program as requirements or electives, and updates the `create_courseware` command to support that same tagging.  

#### How should this be manually tested?

Automated tests should pass.

Testing the `create_courseware` management command: See the documentation for details. 
* The `--required` and `--elective` flags are new and will flag the course appropriately within the program. 
* The command will now error if you swap the readable ID and title fields.
* It will also ignore `--required` and `--elective` if the course you're creating won't be live. (Having a not-live course in the requirements tree causes Django Admin to have issues.) 

Testing the program drawer: 
* You will need to either turn the feature flag for the program UI on or use the `?enable_programs` flag on the URI to see the program drawer.
* The tags should display accordingly for the status of the course.
* Opening the program drawer should present a UI that looks like the InVision layout in desktop mode.
* The program drawer should span the width of the screen on a mobile device (< 1024px horizontally). 
* The program drawer should list the total number of courses in the program, the number you've passed, and the number you've not enrolled in.
* The program drawer should list all live courses that are represented in the program's requirements tree. (In other words, it _will not_ display a course that isn't an elective or a requirement.) 
* The program drawer should group the courses by their status (elective or requirement). The groupings should show a count of programs in that particular status.
* The program drawer should show the standard EnrolledItemCard for courses that you have an enrollment in, and the ProgramCourseInfoCard for ones that you don't. 
* You should see upsell messaging for courses that you're not enrolled in.
   * The Enroll button on the enrollment specific upsell message should take you to the course details page. 
   * The Financial Aid? link on the enrollment specific upsell message should take you to the configured finaid form for the course (this may be the program's form, and will depend on how you have that set up)
* Other functionality should function as per usual.

Accessibility: The Program Drawer should gain focus when opened and the individual cards should be read in order while using a screen reader. (We are reusing elements here that worked OK for accessibility so the page should work as expected using accessible means.) 

#### Screenshots (if appropriate)
My Courses dashboard (with new tags):
![image](https://user-images.githubusercontent.com/945611/200673381-82c459bb-b2e7-4693-9ed8-f821bfadc966.png)

Program Drawer (open): 
![image](https://user-images.githubusercontent.com/945611/200673463-b8a67b26-183e-4ee0-b2e1-e9172cb2d6f5.png)

Program Drawer (open, phone):
![image](https://user-images.githubusercontent.com/945611/200674080-35f2e39e-bb27-49eb-9a52-097ff59ef9f4.png)

Program Drawer (open, tablet):
![image](https://user-images.githubusercontent.com/945611/200673633-653e8696-2862-4bae-8489-186f537e09d5.png)

collinpreston commented on frontend/public/src/components/EnrolledItemCard.js:
```diff
@@ -344,8 +356,29 @@ export class EnrolledItemCard extends React.Component<
+                      Enrolled in certificate track
+                    </span>
+                  ) : null}
+                  {startDateDescription !== null ? (
```
Does `startDateDescription` always have a defined value as long as a start_date is set for the course_run?

collinpreston commented on frontend/public/src/components/ProgramEnrollmentDrawer.js:
```diff
@@ -28,10 +31,58 @@ export class ProgramEnrollmentDrawer extends React.Component<ProgramEnrollmentDr
+  isRequired(course) {
+    const { enrollment } = this.props
+
+    console.log(
+      `retval will be ${enrollment.program.requirements.required.indexOf(
+        course.id
+      )}`
+    )
```
Were these console.log's (4 of them) for your debugging or intended for production?

collinpreston reviewed (commented):
Added some comments.  Still need to functionally test today.

jkachel commented on courses/serializers.py:
```diff
@@ -334,6 +335,23 @@ def get_topics(self, instance):
         )
         return list(topics)
 
+    def get_requirements(self, instance):
```
Hmm. Not sure - I named it requirements because the model is `ProgramRequirement`. You have to satisfy the electives to be able to complete the program too, though, I thought.

collinpreston replied:
Your explanation makes sense to me.

jkachel commented:
Fixed the low-hanging fruit items in https://github.com/mitodl/mitxonline/pull/1198/commits/56cd97fcaed1db19ac3b99c9ba99b63b2939fe28 and rebased from main.

collinpreston commented on frontend/public/src/components/ProgramCourseInfoCard.js:
```diff
@@ -70,7 +81,32 @@ export class ProgramCourseInfoCard extends React.Component<ProgramCourseInfoCard
+              <div className=" w-100 ">
+                <strong>Enroll today</strong> for free, and start learning from
+                MIT faculty. Courses are always free until you want to get a
+                certtificate.
```
certtificate -> certificate

collinpreston reviewed (commented):
One the iPhone SE size screen, the course tags are going off the screen.
<img width="365" alt="Screen Shot 2022-11-10 at 11 24 26" src="https://user-images.githubusercontent.com/8311573/201151034-48e69a2a-ccbe-41da-a1db-cef8201f3537.png">

collinpreston reviewed (commented):
Never mind I read your requirements again!  If a course belongs to a program, will the course always either be an elective or a pre-req?

collinpreston reviewed (approved):
Another big PR!  Looks good!
